### DISCUSSION TITLE
Weird permission requests

### DISCUSSION THREAD

**Discussion by KaKi87** [denoland/deno#25471] [Q&A]
Hi,

I'm encountering weird permission requests when running my [`main.js`](https://git.kaki87.net/KaKi87/dynapt/src/commit/af3fa0a1e6467e400505e71a1d5bc735ef43ae98/main.js) without `-A` :

```
⚠️  Deno requests env access to "NODE_V8_COVERAGE".
⚠️  Deno requests env access to "V8_COVERAGE".
⚠️  Deno requests sys access to "hostname".
⚠️  Deno requests read access to <main_module>.
⚠️  Deno requests read access to "/home/kaki/.cache/deno/node_modules".
```

Why ?

Thanksd

**Comment by bartlomieju** [comment]
There's nothing weird about it, you are using `npm:commander` package, it's that package that requests these permissions.

**Reply by KaKi87** [reply]
Well, I thought so, but then I looked for [`NODE_V8_COVERAGE` at tj/commander.js](https://github.com/search?q=repo%3Atj%2Fcommander.js%20NODE_V8_COVERAGE&type=code), which returns nothing, same for the other environment variable ;

Also I can't figure out what *sys access to "hostname"* and *read access to <main_module>* mean ;

Finally, *Deno requests read access to "/home/kaki/.cache/deno/node_modules"* is plain weird, like as if `import` statements required permissions, which doesn't sound right, but if I reject anyway then the script still works, and if I try `--allow-read="/home/kaki/.cache/deno/node_modules"` then it still prompts for it...

**Reply by bartlomieju** [reply]
Try executing just this line: `import createFastify from 'npm:fastify@^4';`
```
deno run main.js
✅ Granted env access to "NODE_V8_COVERAGE".
✅ Granted env access to "V8_COVERAGE".
✅ Granted sys access to "hostname".
```

These three are getting asked by `fastify` (or its dependencies).

The `<main_module>` permissions is queried if you use `Deno.mainModule` or `process.argv`.

> Finally, Deno requests read access to "/home/kaki/.cache/deno/node_modules" is plain weird, like as if import statements required permissions, which doesn't sound right, but if I reject anyway then the script still works, and if I try --allow-read="/home/kaki/.cache/deno/node_modules" then it still prompts for it...

No, these permissions are only asked by `require()` calls - because they accept any JS expression they can't be statically analyzed and we need to query for permissions there. Most likely the `require()` call is wrapped in `try {} catch {}` and that's why the program still works. Some libraries do that to try and load a certain library (eg. `v8-code-cache` or `source-maps)`.

Hope that helps.

**Reply by 0f-0b** [reply]
`fastify` uses `pino` for logging. `pino` [includes the hostname in the log messages](https://github.com/pinojs/pino/tree/v9.4.0#usage). `pino` also accesses the [`V8_COVERAGE` environment variables](https://github.com/pinojs/pino/blob/v9.4.0/lib/tools.js#L234) to work around [an issue in older versions of Node.js](https://github.com/nodejs/node/issues/49344).

`commander` reads `process.argv[1]` in advance in case it needs to [infer the program name](https://github.com/tj/commander.js/tree/v12.1.0#name) or to search for [standalone executable subcommands](https://github.com/tj/commander.js/tree/v12.1.0#stand-alone-executable-subcommands). Since you don't use either feature, you can prevent `commander` from doing this by changing the last line of `main.js` to the following.

```js
program.parse(Deno.args, { from: 'user' });

// or if you don't want to use a Deno-specific API
import process from 'node:process';
program.parse(process.argv.slice(2), { from: 'user' });
```

**Reply by KaKi87** [reply]
Wow, thank you two very much for such a thorough analysis !

> No, these permissions are only asked by `require()` calls

Okay, how do I get those granted though ? Initially I tried `--allow-read="~/.cache/deno/node_modules"` without success, then I tried `--allow-read="$HOME/.cache/deno/node_modules"` still without success, so I put the exact path `--allow-read="/home/kaki/.cache/deno/node_modules"` but even that didn't work :/

> `fastify` uses `pino` for logging. `pino` [includes the hostname in the log messages](https://github.com/pinojs/pino/tree/v9.4.0#usage). `pino` also accesses the [`V8_COVERAGE` environment variables](https://github.com/pinojs/pino/blob/v9.4.0/lib/tools.js#L234) to work around [an issue in older versions of Node.js](https://github.com/nodejs/node/issues/49344).

I see, but the logger is also [disabled by default](https://fastify.dev/docs/latest/Reference/Server/#logger) :thinking:

**Reply by JhonnyJason** [reply]
This is an interesting issue, I also encountered it with some other package I import from the npm registry.
What I don't get is why I never had encountered this before while I was using npm modules many times, I would assume that deno would use the registry cache all of the time.

So from what I read here I conclude: The reason for the permission requests are the require-calls inside the npm package. This might be because usually dependencies are gathered by deno before actually running deno with the specific restricted permissions. But on require it is done at runtime when the permissions are already restricted. 

- Is this correct?
- What is the the good way to fix it?
    - always adding -A in this case? (Though I want most exact permissions)
    - specifying different permissions depending on which packages I use? (Though I don't want to upgrade my tasks with these things whenever I change my dependencies, as changing dependencies happens fast in development - this would be a chronic pain )
    - Update all my dependenceis to not use require? (Phew, this might be a lot of work but maybe worth it in some cases)
    
    
I think from it's nature it is not possible to change this behavior from within deno. Or well "possible" it might be a bad design choice as generally require behaves like an run-time read. You may write out arbitrary JS files and then parse them via require - okay wait, how are dynamic imports much different?...

**Reply by bartlomieju** [reply]
> You may write out arbitrary JS files and then parse them via require - okay wait, how are dynamic imports much different?...

`import ...` and `import()` are statically analyzable (with limits like `import("" + `${var}`) compared to `require()` which can be aliased to a single letter variable or something else that's obscure.

**Reply by JhonnyJason** [reply]
For now my middle-ground-fix is to simply set the general `--allow-read` - for development that is fine I would say ;-)
